POSTS: Create general post comments

robin-macpherson opened the pull request:
## Description

**Issue:** 

- closes #278 
- completes 2/3 of #210 

Users should also be able to leave general comments on a Post.

- Creates new models: `PostComment` & `PostCommentLike` with migration
- Cleans up and organizes `nexus/*` and `graphql/*` files
- Also creates the `deleteThreadMutation` that will be used in a follow-up PR to handle automatic empty thread deletion (this was going to be included in this PR but the PR got quite large so moving the remaining work to a separate PR to try and keep this one manageable for code review)

## Code Review Notes

- This PR includes finally organizing the `nexus/*` and `graphql/*` files. It's quite clear in the diff for the most part when files are just being moved, but I'm leaving comments where it might be helpful

## Subtasks

- [x] I have added this PR to the Journaly Kanban project ✅
- [x] Create new models
- [x] Update GraphQL schema
- [x] Update GQL hooks
- [x] Finally clean up / organize `graphql/*`
- [x] Hook up the UI!
- [x] Make it look fairly noice!
- [x] Prep us for automatic thread deletion in follow-up PR

## Post Comment Flow
![post-comments-preview](https://user-images.githubusercontent.com/34203886/90303777-c831ee00-de65-11ea-8269-3bb7e3a902f7.gif)

robin-macpherson commented:
This is currently working with new comments as well as editing and deleting, and looks quite ok although I'll want to make it look nicer very soon. Working on some cleanup right now and hooking up automatic thread deletion

robin-macpherson commented:
Build errors fixed ✅

robin-macpherson commented on nexus/comment.ts:
```diff
@@ -0,0 +1,384 @@
+import { schema } from 'nexus'
```
A lot of things in this file were moved from the general `graphql.ts` file. I'll comment to indicate the new stuff

robin-macpherson commented on nexus/comment.ts:
```diff
@@ -0,0 +1,384 @@
+    t.model.body()
+    t.model.createdAt()
+  },
+})
```
The `PostComment` object type is new (30 - 39)

robin-macpherson commented on nexus/comment.ts:
```diff
@@ -0,0 +1,384 @@
+          },
+        })
+      },
+    })
```
`deleteThread` is new (76 - 103)

robin-macpherson commented on nexus/comment.ts:
```diff
@@ -0,0 +1,384 @@
+        return comment
+      },
+    })
+
```
Everything below this point in this file is new

robin-macpherson commented on nexus/graphql.ts:
```diff
@@ -1,32 +1,5 @@
 import { schema } from 'nexus'
 
```
All the changes here are just moving things to other files for organization

robin-macpherson commented on nexus/utils.ts:
```diff
@@ -162,10 +154,7 @@ export const processEditorDocument = (document: NodeType[]) => {
-  original: Post | Comment,
-  currentUser: User,
-) => {
+export const hasPostPermissions = (original: Post | Comment | PostComment, currentUser: User) => {
```
The only substantive change in this file is adding the `PostComment` union

Lanny replied:
To avoid having to continue to expand this union like this, we could change its signature like:

```
type AuthoredObject = { authorId: int }
export const hasAuthorPermissions(original: AuthoredObject, currentUser: User) {
...
```

robin-macpherson replied:
Done! Great idea!

Lanny commented on nexus/comment.ts:
```diff
@@ -0,0 +1,384 @@
+        postCommentId: intArg({ required: true }),
+      },
+      resolve: async (_parent, args, ctx) => {
+        const { userId } = ctx.request
+        if (!userId) throw new Error('You must be logged in to do that.')
+
+        const currentUser = await ctx.db.user.findOne({
+          where: {
+            id: userId,
+          },
+        })
+
+        if (!currentUser) throw new Error('User not found.')
```
We do this block of logic in a bunch of places. It would be nice to have a util function like `getLoggedInUser(ctx)` that returns the current user or throws if they're not logged in or can't be found in the DB

robin-macpherson replied:
Scheduled for a follow-up! ✅

Lanny commented on pages/post/[id]/index.tsx:
```diff
@@ -19,15 +23,28 @@ const PostPage: NextPage = () => {
   })
   const { loading: userLoading, error: userError, data: userData } = useCurrentUserQuery()
 
+  const post: PostType | undefined | null = postData?.postById
```
I'm definitely happy to see the null check lifted out of the `Post` component, maybe I missed it but did we update `Post` to not allow `null` or `undefined` values for the `post` prop, and remove null checking around that?

robin-macpherson replied:
DONE! ✅

Lanny commented on prisma/migrations/migrate.lock:
```diff
@@ -6,4 +6,6 @@
-20200805064655-add-published-at-to-posts
\ No newline at end of file
+20200805064655-add-published-at-to-posts
+20200813085658-add-post-comment
+20200813092954-add-post-comment
```
I don't see `20200813085658-add-post-comment` directory, does that mig need to be removed from this file?

robin-macpherson replied:
Great catch! ✅

Lanny commented on prisma/schema.prisma:
```diff
@@ -73,27 +75,28 @@ model SocialMedia {
+  latitude     Float?
+  threads      Thread[]
+  postTopics   PostTopic[]
+  postComments PostComment[]
```
heh, this diff is maybe a good argument for why the cute alignment thing the prisma tooling enforces isn't such a great idea. No worries though.

robin-macpherson replied:
Haha agreed!

robin-macpherson commented on nexus/comment.ts:
```diff
@@ -0,0 +1,384 @@
+  definition(t) {
+    t.model.id()
+    t.model.author()
+    t.model.authorId()
```
✅

robin-macpherson commented on components/Dashboard/Post/PostComment.tsx:
```diff
@@ -0,0 +1,162 @@
+          display: flex;
+        }
+
+        :global(.edit-btn) {
```
Great point! Done ✅

robin-macpherson commented on components/Dashboard/Post/PostComment.tsx:
```diff
@@ -0,0 +1,162 @@
+      </div>
+      <div className="body-block">
+        {isEditMode ? (
+          <textarea
```
Great point, too! I think what I will do for now is just add `resize: vertical;` manually and handle the `.j-textarea` in my design review PR today so avoid any unexpected stuff in this pretty large PR!

robin-macpherson commented on components/Dashboard/Post/PostComments.tsx:
```diff
@@ -1,18 +1,95 @@
+      </div>
+      {currentUser && (
+        <form onSubmit={createNewPostComment}>
+          <fieldset>
```
Done ✅

robin-macpherson commented:
@nlitwin @Lanny addressed all feedback, thanks for the great reviews! 🤩 I'm going to run through one more round of staging testing and then will apply the migration to the production DB and merge this baby in 🚼 🚀
